File: 14-1.rb
#!/usr/bin/env ruby

program = File.read('14.input').lines.map(&:strip)

memory = {}
mask = 'X' * 36

instruction_format = /(?<inst>mask|mem)(\[(?<addr>[[:digit:]]+)\])? = (?<oper>[[:digit:]X]+)/

def apply_mask(value, mask)
  masked = 0
  0.upto 35 do |exp|
    case mask[35 - exp]
    when '0'
    when '1'
      masked += 2**exp
    when 'X'
      masked += value & 2**exp
    end
  end
  masked
end

program.each do |insn|
  instruction_format.match(insn) do |m|
    case m['inst']
    when 'mask'
      mask = m['oper']
    when 'mem'
      memory[m['addr'].to_i] = apply_mask(m['oper'].to_i, mask)
    else
      print "Illegal instruction: #{insn}\n"
      exit
    end
  end
end

print memory.values.reduce(0, :+), "\n"

File: 14-1.py
#!/usr/bin/env python

import re
from functools import reduce
import operator

program = [insn.strip() for insn in open('14.input').readlines()]

memory = {}
mask = 'X' * 36

instruction_format = re.compile('(?P<inst>mask|mem)(\[(?P<addr>[0-9]+)\])? = (?P<oper>[0-9X]+)')

def apply_mask(value, mask):
    masked = 0
    for exp in range(36):
        if mask[35 - exp] == '0':
            pass
        elif mask[35 - exp] == '1':
            masked += 2**exp
        elif mask[35 - exp] == 'X':
            masked += value & 2**exp
    return masked

for insn in program:
    m = instruction_format.match(insn)
    if m['inst'] == 'mask':
        mask = m['oper']
    elif m['inst'] == 'mem':
        memory[int(m['addr'])] = apply_mask(int(m['oper']), mask)
    else:
      print("Illegal instruction: {insn}")
      sys.exit

print(f"{reduce(operator.add, memory.values())}")
